### PR TITLE
refactor: conversation event processor decoding [WPB-5040]

### PR DESCRIPTION
**Pull request by markusfassbender**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-5040" title="WPB-5040" target="_blank">Sub-taskWPB-5040</a>  [iOS] Process protocol change event
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Refactor the event payload decoding and processing.

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

It exists automated tests.

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-transport/Source/Public/ZMUpdateEvent.swift [hunk 1]
```diff
@@ -115,7 +115,7 @@ extension ZMUpdateEventType {
         case .conversationRename:
             return "conversation.rename"
         case .conversationProtocolUpdate:
-            return "conversation.protocol-update"
+            return "conversation.protocol-update" // TODO: this is the event to be used
```
Comment (caldrian) [inline wire-ios-transport/Source/Public/ZMUpdateEvent.swift]: What does this TODO mean? 
Can it be fixed before merging?

Comment (markusfassbender) [reply wire-ios-transport/Source/Public/ZMUpdateEvent.swift]: Sorry removed it

File: wire-ios-request-strategy/Tests/Sources/Synchronization/Decoding/EventPayloadDecoderTests.swift [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import XCTest
+@testable import WireRequestStrategy
+
+final class EventPayloadDecoderTests: XCTestCase {
```
Comment (markusfassbender) [inline wire-ios-request-strategy/Tests/Sources/Synchronization/Decoding/EventPayloadDecoderTests.swift:22]: Now it's easy to write unit tests without constructing a `ZMUpdateEvent` in core data first 😉

File: wire-ios-request-strategy/Sources/Synchronization/Decoding/EventPayloadDecoder.swift [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    func decode<T>(
+        _ type: T.Type,
+        from eventPayload: [AnyHashable: Any]
+    ) throws -> T where T: Decodable {
+        let data: Data
+
+        do {
+            data = try JSONSerialization.data(withJSONObject: eventPayload, options: [])
```
Comment (caldrian) [inline wire-ios-request-strategy/Sources/Synchronization/Decoding/EventPayloadDecoder.swift:44]: To me it looks like to be a bit redundant.
The Codable implementation is said to use JSONSerialization under the hood so we are repeating a step here. But it's probably convenient, I assume.

Btw, do we need this in the app code or could it be moved to the test target?

File: wire-ios-request-strategy/Sources/Synchronization/Decoding/EventPayloadDecoder.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return try decode(type, from: data)
+    }
+
+    func decode<T>(
+        _ type: T.Type,
+        from eventPayload: Data
+    ) throws -> T where T: Decodable {
+        do {
+            return try decoder.decode(type, from: eventPayload)
+        } catch let error {
```
Comment (caldrian) [inline wire-ios-request-strategy/Sources/Synchronization/Decoding/EventPayloadDecoder.swift]:
```suggestion
        } catch {
```

I was about to make a search/replace in the whole workspace for this ☺️.
Do you prefer to be explicit? I usually skip any character that is not strictly required.

Comment (markusfassbender) [reply wire-ios-request-strategy/Sources/Synchronization/Decoding/EventPayloadDecoder.swift]: Good catch, I will change it 🤙

**Review by caldrian — APPROVED** [review]

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationEventProcessor.swift [hunk 5]
```diff
@@ -59,116 +60,168 @@ public class ConversationEventProcessor: NSObject, ConversationEventProcessorPro
...
+        default:
+            break
+        }
+    }
+
+    private func processConversationCreate(_ event: ZMUpdateEvent) async {
+        guard let payload = try? eventPayloadDecoder.decode(Payload.ConversationEvent<Payload.Conversation>.self, from: event.payload) else {
+            return
+        }
+        await processor.processPayload(payload, in: context)
```
Comment (David-Henner) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationEventProcessor.swift]: What do you think about having this on multiple lines to make it more readable?
```suggestion
        guard let payload = try? eventPayloadDecoder.decode(
            Payload.ConversationEvent<Payload.Conversation>.self, 
            from: event.payload
        ) else { return }
        await processor.processPayload(payload, in: context)
```

Comment (markusfassbender) [reply wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationEventProcessor.swift]: Thanks, I updated the formatting

**Review by typfel — APPROVED** [review]

**Review by David-Henner — APPROVED** [review]

**Pull request merged**